application/src/main/java/org/thingsboard/server/actors/ruleChain/RuleNodeActorMessageProcessor.java

```diff
@@ -94,9 +102,16 @@ public void stop(TbActorCtx context) {
     }
 
     @Override
-    public void onPartitionChangeMsg(PartitionChangeMsg msg) {
+    public void onPartitionChangeMsg(PartitionChangeMsg msg) throws Exception {
         if (tbNode != null) {
-            tbNode.onPartitionChangeMsg(defaultCtx, msg);
+            if (!isMySingletonNode()) {
```

Comment:
For the rule nodes for external communication with a single connection allowed, two nodes can live for a moment during the rebalancing process. Because repartition is ordered but not synchronized inside the cluster.
To minimize the downside, consistent hashing may help calculate partitions.